cdc/owner/changefeed.go

```diff
@@ -774,158 +784,87 @@ func (c *changefeed) preflightCheck(captures map[model.CaptureID]*model.CaptureI
+				return err
+			}
+			if eventDone {
+				// It has expired.
+				// We should use the latest table names now.
+				c.currentTables = nil
+				// We should remove the DDL event from
+				// the holder if it has been executed.
+				c.ddlHolder.removeMinDDL()
 			}
-			// If the ddlResolvedTs is greater than barrierTs, we should not execute
-			// the DDL job, because the changefeed is not blocked by the DDL job yet,
-			// which also means not all data before the DDL job is sent to downstream.
-			// For example, let say barrierTs(ts=10) and there are some ddl jobs in
-			// the queue: [ddl-1(ts=11), ddl-2(ts=12), ddl-3(ts=13)] => ddlResolvedTs(ts=11)
-			// If ddlResolvedTs(ts=11) > barrierTs(ts=10), it means the last barrier was sent
-			// to sink is barrierTs(ts=10), so the data have been sent ware at most ts=10 not ts=11.
-			c.barriers.Update(ddlJobBarrier, ddlResolvedTs)
-			_, barrierTs = c.barriers.Min()
-			return barrierTs, nil
-		}
-
-		// TiCDC guarantees all dml(s) that happen before a ddl was sent to
-		// downstream when this ddl is sent. So, we need to wait checkpointTs is
-		// fullyBlocked at ddl resolvedTs (equivalent to ddl barrierTs here) before we
-		// execute the next ddl.
-		// For example, let say there are some events are replicated by cdc:
-		// [dml-1(ts=5), dml-2(ts=8), ddl-1(ts=11), ddl-2(ts=12)].
-		// We need to wait `checkpointTs == ddlResolvedTs(ts=11)` before execute ddl-1.
-		if !checkpointReachBarrier {
-			return barrierTs, nil
-		}
-
-		done, err := c.asyncExecDDLJob(ctx, ddlJob)
-		if err != nil {
-			return 0, errors.Trace(err)
-		}
-		if !done {
-			return barrierTs, nil
 		}
+	}
+	// update tableBarrier
+	c.barrier.TableBarrier = c.ddlHolder.getAllTableBarrier()
+	return nil
+}
 
-		// If the last ddl was executed successfully, we can pop it
-		// from ddlPuller and update the ddl barrierTs.
-		c.lastDDLTs = ddlResolvedTs
-		c.ddlPuller.PopFrontDDL()
-		newDDLResolvedTs, _ := c.ddlPuller.FrontDDL()
-		c.barriers.Update(ddlJobBarrier, newDDLResolvedTs)
+func (c *changefeed) handleGlobalBarrier(ctx cdcContext.Context) error {
+	barrierTp, barrierTs := c.barriers.Min()
+	c.metricsChangefeedBarrierTsGauge.Set(float64(oracle.ExtractPhysical(barrierTs)))
+	checkpointReachBarrier := barrierTs == c.state.Status.CheckpointTs
+	switch barrierTp {
 	case syncPointBarrier:
-		if !fullyBlocked {
-			return barrierTs, nil
+		if !checkpointReachBarrier {
+			break
 		}
 		nextSyncPointTs := oracle.GoTimeToTS(oracle.GetTimeFromTS(barrierTs).Add(c.state.Info.Config.SyncPointInterval))
 		if err := c.sink.emitSyncPoint(ctx, barrierTs); err != nil {
-			return 0, errors.Trace(err)
+			return errors.Trace(err)
 		}
 		c.barriers.Update(syncPointBarrier, nextSyncPointTs)
 	case finishBarrier:
-		if fullyBlocked {
+		if checkpointReachBarrier {
 			c.feedStateManager.MarkFinished()
 		}
-		return barrierTs, nil
+		break
+	case ddlResolvedBarrier:
+		c.barriers.Update(ddlResolvedBarrier, c.ddlPuller.ResolvedTs())
```

Comment:
Redo ddlResolvedTs is also a type of ddlResolvedBarrier,  so should we write DDL events to redo in ddl_holder and get ddl-resolvedTs from `ddl_holder`?  
In addition, encapsulating all schema and DDL related logic inside `ddl_holder` would be a better way to improve code maintainability?

Reply:
good idea